src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.cpp

```diff
@@ -14,19 +14,119 @@
 #include "DataStructures/DataVector.hpp"        // IWYU pragma: keep
 #include "DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp"
 #include "DataStructures/Tensor/Tensor.hpp"  // IWYU pragma: keep
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "DataStructures/Variables.hpp"
 #include "Parallel/PupStlCpp11.hpp"
 #include "PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.hpp"
 #include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
 #include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/ContainerHelpers.hpp"
 #include "Utilities/Gsl.hpp"
 #include "Utilities/MakeWithValue.hpp"
 #include "Utilities/StdArrayHelpers.hpp"
 #include "Utilities/StdHelpers.hpp"
+#include "Utilities/TMPL.hpp"
+
+// IWYU pragma: no_forward_declare Tensor
 
 /// \cond
 namespace gr {
 namespace Solutions {
 
+namespace {
+template <typename DataType>
+struct KerrSchildBuffer;
+
+template <>
+struct KerrSchildBuffer<double> {
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init)
+  explicit KerrSchildBuffer(const size_t /*size*/) noexcept {}
+
+  tnsr::I<double, 3> x_minus_center;
+  double a_dot_x;
+  double a_dot_x_squared;
+  double half_xsq_minus_asq;
+  double r_squared;
+  double a_dot_x_over_rsquared;
+  double deriv_log_r_denom;
+  tnsr::i<double, 3> deriv_log_r;
+  double H_denom;
+  double temp1;
+  double temp2;
+  tnsr::i<double, 3> a_cross_x;
+  double denom;
+  double r;
+
+  double H;
+  tnsr::i<double, 3> deriv_H;
+  tnsr::i<double, 3> null_form;
+  tnsr::ij<double, 3> deriv_null_form;
+  double null_vector_0 = -1.0;
+  double lapse_squared;
+};
+
+template <>
+struct KerrSchildBuffer<DataVector> {
+  private:
+   // We make one giant allocation so that we don't thrash the heap. This is
+   // important if we call Kerr-Schild frequently such as during ray tracing or
+   // applying analytic boundary conditions.
```

Comment:
I appreciate this comment, but it bothers me to justify the DataVector specialization with ray tracing as an argument, because ray tracing is much more likely to use the double specialization. Perhaps remove the ray-tracing argument altogether? Or move this comment somewhere that it applies to both the double and DataVector cases.